src/parser/parser_verification.js

```diff
@@ -0,0 +1,180 @@
+ */
+function verifyRequiredValues(node, emit, isAdInline) {
+  if (!node || !node.nodeName) {
+    return;
+  }
+  if (node.nodeName === 'InLine') {
+    isAdInline = true;
+  }
+  verifyRequiredAttributes(node, emit);
+
+  if (hasSubElements(node)) {
+    verifyRequiredSubElements(node, emit, isAdInline);
+    for (const nodeKey in node.children) {
+      verifyRequiredValues(node.children[nodeKey], emit, isAdInline);
+    }
+  } else if (parserUtils.parseNodeText(node).length === 0) {
+    emitMissingValueWarning(
+      { name: node.nodeName, parentName: node.parentNode.nodeName },
+      emit
+    );
+    return;
+  }
+}
+
+/**
+ * Verify and trigger warnings if node required attributes are not set.
+ * @param  {Node} node - The node element.
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ */
+function verifyRequiredAttributes(node, emit) {
+  if (
+    !requiredValues[node.nodeName] ||
+    !requiredValues[node.nodeName].attributes
+  ) {
+    return;
+  }
+  const requiredAttributes = requiredValues[node.nodeName].attributes;
+  const missingAttributes = [];
+  requiredAttributes.forEach(attributeName => {
+    const attribute = node.getAttribute(attributeName);
+    if (!attribute) {
+      missingAttributes.push(attributeName);
+    }
+  });
+  if (missingAttributes.length > 0) {
+    emitMissingValueWarning(
+      {
+        name: node.nodeName,
+        parentName: node.parentNode.nodeName,
+        attributes: missingAttributes
+      },
+      emit
+    );
+  }
+}
+
+/**
+ * Verify and trigger warnings if node required sub element are not set.
+ * @param  {Node} node - The node element
+ * @param  {Boolean} isAdInline - True if node is contained in a inline
+ * @param  {Function} emit - Emit function used to trigger Warning event.
+ */
+function verifyRequiredSubElements(node, emit, isAdInline) {
+  const required = requiredValues[node.nodeName];
+
+  if (!required) {
+    return;
+  }
+
+  if (required.subElements) {
+    const requiredSubElements = required.subElements;
+    const missingSubElements = [];
+
+    requiredSubElements.forEach(subElementName => {
+      const subElement = parserUtils.childByName(node, subElementName);
+      if (!subElement) {
+        missingSubElements.push(subElementName);
+      }
+    });
+
+    if (missingSubElements.length > 0) {
+      emitMissingValueWarning(
+        {
+          name: node.nodeName,
+          parentName: node.parentNode.nodeName,
+          subElements: missingSubElements
+        },
+        emit
+      );
+    }
+  }
+
+  // When InLine format is used some nodes (i.e <NonLinear>, <Companion>, or <Icon>)
+  // require at least one of the following resources: StaticResource, IFrameResource, HTMLResource
+  if (!isAdInline || !required.inLineResources) {
+    return;
+  }
+
+  const resourceFound = required.inLineResources.some(resource => {
```

Comment:
are all the InLine subelements here optional, like `StaticResource, IFrameResource, HTMLResource`?

Reply:
as discussed in person, we shoud rename `inLineResources` to `oneOfinLineResources`